crates/ty_python_semantic/resources/mdtest/final.md

```diff
@@ -29,3 +29,344 @@ class H(
+
+class ChildOfGood(Good):
+    @overload
+    def bar(self, x: str) -> str: ...
+    @overload
+    def bar(self, x: int) -> int: ...  # error: [override-of-final-method]
+    @overload
+    def baz(self, x: str) -> str: ...
+    @overload
+    def baz(self, x: int) -> int: ...  # error: [override-of-final-method]
+
+class Bad:
+    @overload
+    def bar(self, x: str) -> str: ...
+    @overload
+    @final
+    # error: [invalid-overload]
+    def bar(self, x: int) -> int: ...
+    @overload
+    def baz(self, x: str) -> str: ...
+    @final
+    @overload
+    # error: [invalid-overload]
+    def baz(self, x: int) -> int: ...
+
+class ChildOfBad(Bad):
+    @overload
+    def bar(self, x: str) -> str: ...
+    @overload
+    def bar(self, x: int) -> int: ...  # error: [override-of-final-method]
+    @overload
+    def baz(self, x: str) -> str: ...
+    @overload
+    def baz(self, x: int) -> int: ...  # error: [override-of-final-method]
+```
+
+`main.py`:
+
+```py
+from typing import overload, final
+
+class Good:
+    @overload
+    def f(self, x: str) -> str: ...
+    @overload
+    def f(self, x: int) -> int: ...
+    @final
+    def f(self, x: int | str) -> int | str:
+        return x
+
+class ChildOfGood(Good):
+    @overload
+    def f(self, x: str) -> str: ...
+    @overload
+    def f(self, x: int) -> int: ...
+    # error: [override-of-final-method]
+    def f(self, x: int | str) -> int | str:
+        return x
+
+class Bad:
+    @overload
+    @final
+    def f(self, x: str) -> str: ...
+    @overload
+    def f(self, x: int) -> int: ...
+    # error: [invalid-overload]
+    def f(self, x: int | str) -> int | str:
+        return x
+
+    @final
+    @overload
+    def g(self, x: str) -> str: ...
+    @overload
+    def g(self, x: int) -> int: ...
+    # error: [invalid-overload]
+    def g(self, x: int | str) -> int | str:
+        return x
+
+    @overload
+    def h(self, x: str) -> str: ...
+    @overload
+    @final
+    def h(self, x: int) -> int: ...
+    # error: [invalid-overload]
+    def h(self, x: int | str) -> int | str:
+        return x
+
+    @overload
+    def i(self, x: str) -> str: ...
+    @final
+    @overload
+    def i(self, x: int) -> int: ...
+    # error: [invalid-overload]
+    def i(self, x: int | str) -> int | str:
+        return x
+
+class ChildOfBad(Bad):
+    # TODO: these should all cause us to emit Liskov violations as well
+    f = None  # error: [override-of-final-method]
+    g = None  # error: [override-of-final-method]
+    h = None  # error: [override-of-final-method]
+    i = None  # error: [override-of-final-method]
+```
+
+## Edge case: the function is decorated with `@final` but originally defined elsewhere
+
+As of 2025-11-26, pyrefly emits a diagnostic on this, but mypy and pyright do not:
+
+```py
+from typing import final
+
+class A:
+    @final
+    def method(self): ...
+
+class B:
+    method = A.method
+
+class C(B):
+    def method(self): ...  # no diagnostic
```

Comment:
The `no diagnostic` suggests to me that we intentionally assert that we don't raise a diagnsotic here. Is that our intention? If so, why?